### ISSUE TITLE
Unable to make time-series charts in Apache Superset

### ISSUE THREAD

**Issue by akki** [preset-io/elasticsearch-dbapi#20]
Hi

I am trying to build time-series visualisation in Apache Superset but unable to do so.

Below is the screenshot of the error that I get.
<img width="1571" alt="Explore_-_Superset_Admin-Query_public_children_test-FdtlPJnQH" src="https://user-images.githubusercontent.com/4801628/76999546-612d0d80-6989-11ea-8a0a-b7501116f6e0.png">

Here is the stack trace of the error that I get:
```
superset_1         | ERROR:superset.viz:Datetime column not provided as part table configuration and is required by this type of chart
superset_1         | Traceback (most recent call last):
superset_1         |   File "/app/superset/viz.py", line 448, in get_df_payload
superset_1         |     df = self.get_df(query_obj)
superset_1         |   File "/app/superset/viz.py", line 233, in get_df
superset_1         |     self.results = self.datasource.query(query_obj)
superset_1         |   File "/app/superset/connectors/sqla/models.py", line 1034, in query
superset_1         |     query_str_ext = self.get_query_str_extended(query_obj)
superset_1         |   File "/app/superset/connectors/sqla/models.py", line 627, in get_query_str_extended
superset_1         |     sqlaq = self.get_sqla_query(**query_obj)
superset_1         |   File "/app/superset/connectors/sqla/models.py", line 751, in get_sqla_query
superset_1         |     "Datetime column not provided as part table configuration "
superset_1         | Exception: Datetime column not provided as part table configuration and is required by this type of chart
superset_1         | 172.27.0.1 - - [18/Mar/2020 19:26:58] "POST /superset/explore_json/ HTTP/1.1" 400 -
```

On a bit of investigation, I found that the error is because the parameter `granularity` and `granularity_sqla` is missing in the form data being passed to Superset which is [required here](https://github.com/apache/incubator-superset/blob/104f3d25026c23b8f41da4c959da3f59333759c8/superset/viz.py#L302).

I would appreciate if anyone could help out with this issue.

Please let me know if you need any other details from me to help investigate & solve this issue or if you believe that I should be reporting this to [Apache Superset](https://github.com/apache/incubator-superset/issues) instead.

**Comment by villebro** [comment]
@akki can you post a screenshot of your datasource schema? I'm specifically interested in what types of temporal columns you have (`DATETIME` or similar).

**Comment by akki** [comment]
Hi @villebro Thanks for looking into this.


Probably a very novice question but can you please tell me how I can find the "datasource schema"?
Right now, what I can see is that the "Time Column" dropdown shows me 0 options (also can be seen in the pic I shared).

**Comment by akki** [comment]
Also, not sure if this is related but I am not able to create the table from the `/tablemodelview/add` webpage UI.

I get the following error:
```
Table [XXXX] could not be found, please double check your database connection, schema, and table name, error: TransportError(500, 'json_parse_exception', "Unrecognized token 'XXXX': was expecting ('true', 'false' or 'null')\n at [Source: org.elasticsearch.transport.netty4.ByteBufStreamInput@2e7f20e6; line: 1, column: 17]")
```
where I have changed my table name with `XXXX`.

The full stack trace is as follows:
```
superset_1         | DEBUG:elasticsearch:< {"error":{"root_cause":[{"type":"json_parse_exception","reason":"Unrecognized token 'XXXX': was expecting ('true', 'false' or 'null')\n at [Source: org.elasticsearch.transport.netty4.ByteBufStreamInput@2e7f20e6; line: 1, column: 17]"}],"type":"json_parse_exception","reason":"Unrecognized token 'XXXX': was expecting ('true', 'false' or 'null')\n at [Source: org.elasticsearch.transport.netty4.ByteBufStreamInput@2e7f20e6; line: 1, column: 17]"},"status":500}
superset_1         | ERROR:superset.connectors.sqla.views:Got an error in pre_add for XXXX
superset_1         | Traceback (most recent call last):
superset_1         |   File "/app/superset/connectors/sqla/views.py", line 389, in pre_add
superset_1         |     table.get_sqla_table_object()
superset_1         |   File "/app/superset/connectors/sqla/models.py", line 1074, in get_sqla_table_object
superset_1         |     return self.database.get_table(self.table_name, schema=self.schema)
superset_1         |   File "/app/superset/models/core.py", line 585, in get_table
superset_1         |     autoload_with=self.get_sqla_engine(),
superset_1         |   File "<string>", line 2, in __new__
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/deprecations.py", line 128, in warned
superset_1         |     return fn(*args, **kwargs)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/schema.py", line 506, in __new__
superset_1         |     metadata._remove_table(name, schema)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
superset_1         |     compat.reraise(exc_type, exc_value, exc_tb)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 153, in reraise
superset_1         |     raise value
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/schema.py", line 501, in __new__
superset_1         |     table._init(name, metadata, *args, **kw)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/schema.py", line 595, in _init
superset_1         |     resolve_fks=resolve_fks,
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/schema.py", line 619, in _autoload
superset_1         |     _extend_on=_extend_on,
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 2163, in run_callable
superset_1         |     return conn.run_callable(callable_, *args, **kwargs)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1615, in run_callable
superset_1         |     return callable_(self, *args, **kwargs)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 460, in reflecttable
superset_1         |     table, include_columns, exclude_columns, resolve_fks, **opts
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/reflection.py", line 642, in reflecttable
superset_1         |     table_name, schema, **table.dialect_kwargs
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/reflection.py", line 369, in get_columns
superset_1         |     self.bind, table_name, schema, info_cache=self.info_cache, **kw
superset_1         |   File "/elasticsearch-dbapi/es/basesqlalchemy.py", line 163, in get_columns
superset_1         |     f"SHOW ARRAY_COLUMNS FROM {table_name}"
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 976, in execute
superset_1         |     return self._execute_text(object_, multiparams, params)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1149, in _execute_text
superset_1         |     parameters,
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1250, in _execute_context
superset_1         |     e, statement, parameters, cursor, context
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1478, in _handle_dbapi_exception
superset_1         |     util.reraise(*exc_info)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 153, in reraise
superset_1         |     raise value
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1246, in _execute_context
superset_1         |     cursor, statement, parameters, context
superset_1         |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 581, in do_execute
superset_1         |     cursor.execute(statement, parameters)
superset_1         |   File "/elasticsearch-dbapi/es/baseapi.py", line 25, in g
superset_1         |     return f(self, *args, **kwargs)
superset_1         |   File "/elasticsearch-dbapi/es/elastic/api.py", line 136, in execute
superset_1         |     return self.get_array_type_columns(re_table_name[1])
superset_1         |   File "/elasticsearch-dbapi/es/elastic/api.py", line 159, in get_array_type_columns
superset_1         |     resp = self.es.search(table_name, size=1)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/client/utils.py", line 84, in _wrapped
superset_1         |     return func(*args, params=params, **kwargs)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/client/__init__.py", line 1548, in search
superset_1         |     "GET", _make_path(index, doc_type, "_search"), params=params, body=body
superset_1         |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/transport.py", line 358, in perform_request
superset_1         |     timeout=timeout,
superset_1         |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/connection/http_urllib3.py", line 261, in perform_request
superset_1         |     self._raise_error(response.status, raw_data)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/connection/base.py", line 182, in _raise_error
superset_1         |     status_code, error_message, additional_info
superset_1         | elasticsearch.exceptions.TransportError: TransportError(500, 'json_parse_exception', "Unrecognized token 'XXXX': was expecting ('true', 'false' or 'null')\n at [Source: org.elasticsearch.transport.netty4.ByteBufStreamInput@2e7f20e6; line: 1, column: 17]")
```

**Comment by villebro** [comment]
@akki  click on the datasource, choose "Edit datasource" and click "Columns". There you should see at least one column with "Is Temporal" checked.

With regards to adding the table, are you sure you've specified the correct schema and table name exactly like they are displayed under SQL Lab?

**Comment by dpgaspar** [comment]
@akki this may be related to #19, downgrade your `elasticsearch` python pkg to 7.1.0 and retest.

**Comment by akki** [comment]
Hi @villebro & @dpgaspar Thank you for your suggestions.

I was able to mark my field as "Is Temporal" by using the "Edit datasource" feature.

I got another error after that - "search_phase_execution_exception): all shards failed".
Full stacktrace:
```
superset_1         | ERROR:superset.connectors.sqla.models:Query SELECT HISTOGRAM(timestamp, INTERVAL 1 MINUTE) AS __timestamp,
superset_1         |        AVG(prediction) AS "AVG(prediction)",
superset_1         |        AVG("hour_avg_LMTD") AS "AVG(hour_avg_LMTD)"
superset_1         | FROM
superset_1         |   (SELECT *
superset_1         |    FROM XXXX) AS expr_qry
superset_1         | WHERE timestamp >= '2013-03-24 00:00:00.000000'
superset_1         |   AND timestamp < '2020-03-24 00:00:00.000000'
superset_1         | GROUP BY __timestamp
superset_1         | ORDER BY "AVG(prediction)" ASC
superset_1         | LIMIT 10000 on schema None failed
superset_1         | Traceback (most recent call last):
superset_1         |   File "/elasticsearch-dbapi/es/baseapi.py", line 217, in elastic_query
superset_1         |     resp = self.es.transport.perform_request("POST", path, body=payload)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/transport.py", line 358, in perform_request
superset_1         |     timeout=timeout,
superset_1         |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/connection/http_urllib3.py", line 261, in perform_request
superset_1         |     self._raise_error(response.status, raw_data)
superset_1         |   File "/usr/local/lib/python3.6/site-packages/elasticsearch/connection/base.py", line 182, in _raise_error
superset_1         |     status_code, error_message, additional_info
superset_1         | elasticsearch.exceptions.RequestError: RequestError(400, 'search_phase_execution_exception', 'failed to parse date field [2013-03-24 00:00:00.000000] with format [dd-MM-yyyy HH:mm:ss||yyyy-MM-dd HH:mm:ss||yyyy-MM-dd||epoch_millis]: [failed to parse date field [2013-03-24 00:00:00.000000] with format [dd-MM-yyyy HH:mm:ss||yyyy-MM-dd HH:mm:ss||yyyy-MM-dd||epoch_millis]]')
superset_1         |
superset_1         | During handling of the above exception, another exception occurred:
superset_1         |
superset_1         | Traceback (most recent call last):
superset_1         |   File "/app/superset/connectors/sqla/models.py", line 1057, in query
superset_1         |     df = self.database.get_df(sql, self.schema, mutator)
superset_1         |   File "/app/superset/models/core.py", line 360, in get_df
superset_1         |     self.db_engine_spec.execute(cursor, sqls[-1])
superset_1         |   File "/app/superset/db_engine_specs/base.py", line 836, in execute
superset_1         |     cursor.execute(query)
superset_1         |   File "/elasticsearch-dbapi/es/baseapi.py", line 25, in g
superset_1         |     return f(self, *args, **kwargs)
superset_1         |   File "/elasticsearch-dbapi/es/elastic/api.py", line 139, in execute
superset_1         |     results = self.elastic_query(query)
superset_1         |   File "/elasticsearch-dbapi/es/baseapi.py", line 224, in elastic_query
superset_1         |     f"Error ({e.error}): {e.info['error']['reason']}"
superset_1         | es.exceptions.ProgrammingError: Error (search_phase_execution_exception): all shards failed
```
Here
`2013-03-24 00:00:00.000000` is the lower bound of the filter that I had applied to timestamp from the Superset UI. It is basically expecting `2013-03-24 00:00:00.000000` in one of these formats `[dd-MM-yyyy HH:mm:ss||yyyy-MM-dd HH:mm:ss||yyyy-MM-dd||epoch_millis]]`. Those formats are the ones that I had specified in the `mappings` while creating my ES index.
My fields are of those formats so I had provided those in the mappings.

It worked for me when I changed [this line](https://github.com/apache/incubator-superset/blob/06e0f04214bd7866328c6714687eaeffafc44f70/superset/connectors/sqla/models.py#L1032) to `sql = query_str_ext.sql.replace('.000000', '')` - which of course is a hacky solution.

So basically, things won't work if the mapping doesn't have a format which supports datetime values like this one - `2013-03-24 00:00:00.000000`.


I'd leave it up to the repo owners to decide if this seems something like they want to fix or it is expected from the user to provide that mapping.

**Comment by villebro** [comment]
@akki can you check the datatype of your `timestamp` column? The fact that you had to manuaöly make it temporal sounds strange.

**Comment by akki** [comment]
This is from the mapping of my index:
```
 "timestamp": {
    "type": "date",
    "format": "dd-MM-yyyy HH:mm:ss||yyyy-MM-dd HH:mm:ss||yyyy-MM-dd||epoch_millis"
 }
```
The `type` is `date`.


If Superset can detect these fields automatically, that would be a good addition.

**Comment by dpgaspar** [comment]
@akki 

Can't you post the output from:
```
POST _sql
{
  "query": "show columns from akkis_index"
}
```

**Comment by akki** [comment]
@dpgaspar Sure, I can. :)

Here is the output:
```
{
    "columns": [
        {
            "name": "column",
            "type": "keyword"
        },
        {
            "name": "type",
            "type": "keyword"
        },
        {
            "name": "mapping",
            "type": "keyword"
        }
    ],
    "rows": [
        [
            "h_id",
            "VARCHAR",
            "text"
        ],
        [
            "h_id.keyword",
            "VARCHAR",
            "keyword"
        ],
        [
            "hour_avg_LMTD",
            "REAL",
            "float"
        ],
        [
            "prediction",
            "REAL",
            "float"
        ],
        [
            "prediction_datetime",
            "TIMESTAMP",
            "datetime"
        ],
        [
            "timestamp",
            "TIMESTAMP",
            "datetime"
        ]
    ]
}
```

**Comment by dpgaspar** [comment]
Thanks @akki, `prediction_datetime` and `timestamp` should be recognized has DATETIME, can you verify on SQLLab they are identified has such? If yes can you recreate the table?

**Comment by akki** [comment]
Can you tell me how I can verify that?

Also, I would like to re-mention the fact that **I am not able to create the table from the /tablemodelview/add webpage UI** if that was missed earlier.

**Comment by dpgaspar** [comment]
The table problem, should be fixed by downgrading elasticsearch (python pkg) to 7.1.0, or by upgrading this pkg to 0.1.3, can you confirm that?

**Comment by akki** [comment]
Yeah, I don't get that error when using the current latest version of `elasticsearch-dbapi` installed with pip. And now it also auto-detected the `datetime` fields.

So, now I can create my time-series chart, given that I explicitly create a mapping with allowing this format for my datetime fields `yyyy-MM-dd'T'hh:mm:ss'Z'`.

IMO Superset should send things as `epoch_millis` if it really wants to choose a format but I would be happy to close this issue now as Superset supports many more data stores than just Elasticsearch.
